haskell-best-practices.md

```diff
@@ -13,6 +13,67 @@
      x `divEx` d = x / d
  ```
 
+### Avoid `MonadFail`
+
+Don't write our own functions using `MonadFail`/`fail`.
+
+**Bad**:
+
+```hs
+data Email = Email
+
+parseEmail :: MonadFail m => Text -> m Email
+parseEmail = undefined
+
+instance FromJSON Email where
+  parseJSON = withText "Email" parseEmail
+```
+
+This looks convenient:
+
+```hs
+parseEmail "..." :: Maybe Email -- fail as Nothing
+parseEmail "..." :: IO Email -- fail as throw
+```
+
+It would be reasonable to think you can also:
+
+```hs
+parseEmail "..." :: Either String Email -- fail as Left
+```
```

Comment:
How about something like:

`MonadFail` over `Either` is a common foot-gun. We might expect `fail` to return `Either String a`, but instead it throws an `error`. There are many other surprising and problematic instances of `MonadFail` that make this abstraction dangerous. Writing functions that abstract over `MonadFail` opens us up to these surprising and hard to reason about errors.